### PR TITLE
Add keyboard shortcuts for previous/next chapter #377

### LINKED ISSUE #377: Add keyboard shortcuts for previous/next chapter
Not launch blocking.

It would be nice to be able to "page" through the chapters using keyboard shortcuts and not have to scroll to the bottom of the page for the prev/next links.

I was thinking of using either the "p" and "n" keys or "," and "." (< and >). We should also make these shortcuts discoverable somehow, ideally, for both accessibility and convenience.

### PR DESCRIPTION
**Pull request by arsenicraghav**
Navigate through chapters using **p**(previous) **n**(next) and **,**(previous)  **.**(next)

Closes #377

### CODE REVIEW

File: src/config/2019.json [hunk 1]
```diff
@@ -746,6 +746,16 @@
       "avatar_url": "https:\/\/avatars2.githubusercontent.com\/u\/24447527?v=4&s=200",
       "github": "raghuramakrishnan71"
     },
+    "raghav": {
+      "name": "Raghvendra Kumar",
+      "bio":"Tech Enthusiast and Developer",
```
Comment (bazzadp) [inline src/config/2019.json]: At the moment only Author's bios are used (see #350) so other contributors don't get a bio - sorry! 😊

```suggestion
```

File: src/templates/en/2019/base_chapter.html [hunk 2]
```diff
@@ -119,6 +127,7 @@
 
 {% macro render_prevnext() %}
   {% if prev_chapter %}
+test
```
Comment (bazzadp) [inline src/templates/en/2019/base_chapter.html]: This needed?

```suggestion
```

File: src/templates/en/2019/base_chapter.html [hunk 3]
```diff
@@ -119,6 +127,7 @@
...
       <a id="previous-chapter" href="/{{lang}}/{{year}}/{{prev_chapter['title'].lower().replace(' ', '-').replace('/', '')}}">
```
Comment (bazzadp) [inline src/templates/en/2019/base_chapter.html]: Could we add a title to show the shortcuts on hover? And similarly on next button?

```suggestion
      <a title="Previous Chapter (press 'p' or ',')" id="previous-chapter" href="/{{lang}}/{{year}}/{{prev_chapter['title'].lower().replace(' ', '-').replace('/', '')}}">
```

File: src/templates/en/2019/base_chapter.html [hunk 4]
```diff
@@ -32,6 +32,14 @@
   var el = document.getElementById('num_comments');
   el.innerText = comments + ' ' + (comments == 1 ? 'comment' : 'comments');
 });
+document.addEventListener("keypress", function onPress(event) {
+ 	if (event.key === 'p' || event.key === ',') {
```
Comment (bazzadp) [inline src/templates/en/2019/base_chapter.html]: Should we cater for other versions of these keypresses?

```suggestion
 	if (event.key === 'p' || event.key === 'P' || event.key === ',' || event.key === '<') {
```

File: src/templates/en/2019/base_chapter.html [hunk 5]
```diff
@@ -32,6 +32,14 @@
...
+ 		document.getElementById('previous-chapter').click();
+ 	}
+ 	if (event.key === 'n' || event.key === '.') {
```
Comment (bazzadp) [inline src/templates/en/2019/base_chapter.html]: Should we cater for other versions of these keypresses?

```suggestion
 	if (event.key === 'n' || event.key === 'N' || event.key === '.' || event.key === '>') {
```

File: src/templates/en/2019/base_chapter.html (hunk 4 above)
Comment (arsenicraghav) [reply src/templates/en/2019/base_chapter.html]: We could include check for the uppercase but I think we should not include the arrow("<", ">") keys for this as it's a bit counter-intuitive ( cause arrows are also used for vertical and horizontal scrolling).

File: src/templates/en/2019/base_chapter.html [hunk 6]
```diff
@@ -32,6 +32,14 @@
   var el = document.getElementById('num_comments');
   el.innerText = comments + ' ' + (comments == 1 ? 'comment' : 'comments');
 });
+document.addEventListener("keypress", function onPress(event) {
+ 	if (event.key === 'p' || event.key === ',') {
+ 		document.getElementById('previous-chapter').click();
```
Comment (rviscomi) [inline src/templates/en/2019/base_chapter.html]: Need to handle the edge cases for the first and last chapters, for example:

```suggestion
 		var previous = document.getElementById('previous-chapter').click();
 		if (previous) {
 		  previous.click();
 		}
```

File: src/templates/en/2019/base_chapter.html (hunk 4 above)
Comment (rviscomi) [reply src/templates/en/2019/base_chapter.html]: In this case we're only targeting angled brackets, not the actual arrow keys. I think it's ok to handle `Shift` key versions of other supported shortcut keys.

**Comment by bazzadp** [comment]
@arsenicraghav thanks for your work on this and for the fixes you've put. Any thoughts on my `title` suggestions? Not sure if you've missed that as now outdated.

Also just done a merge to master so you'll need to fix conflicts. Gimme a shout if you need a hand with this.

Not sure if the non-Querty keyboard thing is an issue - anyone got any opinions on this?

@rviscomi any thoughts on accessibility concerns? I checked a few common tools (JAWS, NVDA) and these particular keys don't seem mapped to anything by default but not sure how many people add custom keys and what usual practice is to avoid any issues. Quick Google shows nothing.

**Comment by bazzadp** [comment]
Also updated the first comment so it will auto-close the corresponding issue then merged.

**Comment by arsenicraghav** [comment]
> Should it be easy to switch these off in case they override screen-readers use of these keys?

Yeah, event listener for 'keypress' can be turned off. 

 >Also just done a merge to master so you'll need to fix conflicts.

have resolved the conflicts

>thoughts on my title suggestions

have added 'title' attribute on both the anchor tag

> non-Querty keyboard thing is an issue ?

detecting keyboard layout or type using javascript does not seem possible so not sure how to handle that.  

Ready for review again

File: src/templates/en/2019/base_chapter.html [hunk 7]
```diff
@@ -17,6 +17,36 @@
...
+    return;
+  }
+
+  var comments = +r.posts_count - 1;
+  if (isNaN(comments)) {
+    return;
+  }
+  var el = document.getElementById('num_comments');
+  el.innerText = comments + ' ' + (comments == 1 ? 'comment' : 'comments');
+});
```
Comment (bazzadp) [inline src/templates/en/2019/base_chapter.html]: This whole section has been moved to [static/js/chapter.js](https://github.com/HTTPArchive/almanac.httparchive.org/blob/master/src/static/js/chapter.js) so should not be here any longer.

File: src/config/2019.json [hunk 8]
```diff
@@ -746,6 +746,15 @@
       "avatar_url": "https:\/\/avatars2.githubusercontent.com\/u\/24447527?v=4&s=200",
       "github": "raghuramakrishnan71"
     },
+    "raghav": {
+      "name": "Raghvendra Kumar",
+      "teams": [
+        "developers"
+      ],
+      "avatar_url": "https:\/\/avatars3.githubusercontent.com\/u\/9475742?s=200&v=4",
```
Comment (bazzadp) [inline src/config/2019.json]: Minor nit for consistency:

```suggestion
      "avatar_url": "https:\/\/avatars3.githubusercontent.com\/u\/9475742?v=4&s=200",
```

**Pull request merged**